pkg/dflog/log.go

```diff
@@ -20,113 +20,114 @@ import (
+	return func(l *logrus.Logger) error {
+		if debug {
+			l.SetLevel(logrus.DebugLevel)
+		}
+		return nil
 	}
-	logFile.Seek(0, 2)
+}
 
-	// create formatter for default log Logger
-	formatter := &DragonflyFormatter{
-		TimestampFormat: DefaultLogTimeFormat,
-		Sign:            sign,
+func getLumberjack(l *logrus.Logger) *lumberjack.Logger {
+	if logger, ok := l.Out.(*lumberjack.Logger); ok {
+		return logger
 	}
-
-	// set all details in log default logger
-	logrus.SetLevel(logLevel)
-	logrus.SetOutput(logFile)
-	logrus.SetFormatter(formatter)
 	return nil
 }
 
-// InitConsoleLog initializes console logger for process.
-// console log will output the dfget client's log in console/terminal for
-// debugging usage.
-func InitConsoleLog(debug bool, sign string) {
-	formatter := &DragonflyFormatter{
-		TimestampFormat: DefaultLogTimeFormat,
-		Sign:            sign,
-	}
+// WithLogFile sets the logger to output to the given file, with log rotation.
+// If the given file is empty, nothing will be done.
+func WithLogFile(f string) Option {
+	return func(l *logrus.Logger) error {
+		if f == "" {
+			return nil
+		}
 
-	logLevel := logrus.InfoLevel
-	if debug {
-		logLevel = logrus.DebugLevel
-	}
+		if logger := getLumberjack(l); logger == nil {
+			l.SetOutput(&lumberjack.Logger{
+				Filename:   f,
+				MaxSize:    20, // mb
+				MaxBackups: 1,
+			})
+		} else {
+			logger.Filename = f
+		}
 
-	consoleLog := &logrus.Logger{
-		Out:       os.Stdout,
-		Formatter: formatter,
-		Hooks:     make(logrus.LevelHooks),
-		Level:     logLevel,
+		return nil
 	}
-	hook := &ConsoleHook{
-		logger: consoleLog,
-		levels: logrus.AllLevels,
-	}
-	logrus.AddHook(hook)
 }
 
-// CreateLogger creates a Logger.
-func CreateLogger(logPath string, logName string, logLevel string, sign string) (*logrus.Logger, error) {
-	// parse log level
-	level, err := logrus.ParseLevel(logLevel)
-	if err != nil {
-		level = logrus.InfoLevel
+// WithMaxSizeMB sets the max size of log files in MB. If the logger is not configured
+// to use a log file, an error is returned.
+func WithMaxSizeMB(max uint) Option {
+	return func(l *logrus.Logger) error {
+		if logger := getLumberjack(l); logger != nil {
+			logger.MaxSize = int(max)
+			return nil
+		}
+		return errors.Errorf("lumberjack is not configured")
 	}
+}
 
-	// create log file path
-	logFilePath := path.Join(logPath, logName)
-	if err := os.MkdirAll(filepath.Dir(logFilePath), 0755); err != nil {
-		return nil, err
+// WithConsole add a hook to output logs to stdout
+func WithConsole() Option {
+	return func(l *logrus.Logger) error {
+		consoleLog := &logrus.Logger{
+			Out:       os.Stdout,
+			Formatter: l.Formatter,
+			Hooks:     make(logrus.LevelHooks),
+			Level:     l.Level,
+		}
+		hook := &ConsoleHook{
+			logger: consoleLog,
+			levels: logrus.AllLevels,
+		}
+		l.AddHook(hook)
+		return nil
 	}
+}
 
-	// open log file
-	logFile, err := os.OpenFile(logFilePath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
```

Comment:
It's better to specify the `FileMode` as `0644`, all users can read the log file. If not, the created file's mode is relied on the os's `umask`. If the os's `umask` is `027`, then the new file's mode is `0640`, it means only the owner and users in same group can read the file.

Reply:
The version of lumberjack I'm using, [v2.0.0](https://github.com/natefinch/lumberjack/blob/v2.0.0/lumberjack.go#L215) did set the file mode to `0644`. However in [this commit](https://github.com/natefinch/lumberjack/commit/2e8fbeea3f9e456ee5062dfbcbe3ee902a44bace) it changes the default mode to `0600`.

Reply:
OK. I find that the previous code has same problem with this pr. If we want to create a file with permission `0644` exactly, we should explicitly call `syscall.Umask` to set `umask`.